### PR TITLE
[Merged by Bors] - feat(data/list/alist): recursion on alist using insert

### PR DESCRIPTION
**Pull request by vihdzp**
Mathlib4 pair for https://github.com/leanprover-community/mathlib/pull/15434

---

- [x] depends on: #3013

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
This needs the mathlib SHA (which would have been in the mathport file you copied from) in the header.

File: Mathlib/Data/List/AList.lean [hunk 1]
```diff
@@ -339,6 +345,59 @@ theorem toAList_cons (a : α) (b : β a) (xs : List (Sigma β)) :
...
+#align alist.insert_rec AList.insertRec
+
+-- Test that the `induction` tactic works on `insert_rec`.
+example (l : AList β) : True := by induction l using AList.insertRec <;> trivial
+
+@[simp]
+theorem insertRec_empty {C : AList β → Sort _} (H0 : C ∅)
+    (IH : ∀ (a : α) (b : β a) (l : AList β), a ∉ l → C l → C (l.insert a b)) :
+    @insertRec α β _ C H0 IH ∅ = H0 :=
+  by
```
Comment (eric-wieser) [inline Mathlib/Data/List/AList.lean:372]:
```suggestion
    @insertRec α β _ C H0 IH ∅ = H0 := by
```

File: Mathlib/Data/List/AList.lean [hunk 2]
```diff
@@ -339,6 +345,59 @@ theorem toAList_cons (a : α) (b : β a) (xs : List (Sigma β)) :
...
+  change @insertRec α β _ C H0 IH ⟨[], _⟩ = H0
+  rw [insertRec]
+#align alist.insert_rec_empty AList.insertRec_empty
+
+theorem insertRec_insert {C : AList β → Sort _} (H0 : C ∅)
+    (IH : ∀ (a : α) (b : β a) (l : AList β), a ∉ l → C l → C (l.insert a b)) {c : Sigma β}
+    {l : AList β} (h : c.1 ∉ l) :
+    @insertRec α β _ C H0 IH (l.insert c.1 c.2) = IH c.1 c.2 l h (@insertRec α β _ C H0 IH l) :=
+  by
```
Comment (eric-wieser) [inline Mathlib/Data/List/AList.lean:381]:
```suggestion
    @insertRec α β _ C H0 IH (l.insert c.1 c.2) = IH c.1 c.2 l h (@insertRec α β _ C H0 IH l) := by
```

**Review by eric-wieser — APPROVED** [review]
One of these files had an earlier pending forward-port that I've marked as dependent. Otherwise looks good.

bors d+

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3013
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3013~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**